### PR TITLE
Feature/chunked file support

### PR DESCRIPTION
**Pull request by tnextday**
Auto split large file to many chunks, and marge the chunks when download

### CODE REVIEW

**Comment by yanyiwu** [comment]
wow, big pull request!

File: go/operation/chunked_file.go [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+	}
+	cm := ChunkManifest{}
+	if e := json.Unmarshal(buffer, &cm); e != nil {
+		return nil, e
+	}
+	sort.Sort(cm.Chunks)
+	return &cm, nil
+}
+
+func (cm *ChunkManifest) GetData() ([]byte, error) {
```
Comment (chrislusf) [inline go/operation/chunked_file.go]: How about just use Marshal() here?
GetData() seems a cheap getXXX() operation.

File: go/operation/chunked_file.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+	return json.Marshal(cm)
+}
+
+func (cm *ChunkManifest) DeleteChunks(master string) error {
+	deleteError := 0
+	for _, ci := range cm.Chunks {
+		if e := DeleteFile(master, ci.Fid, ""); e != nil {
+			deleteError++
+			glog.V(0).Infoln("delete error:", e, ci.Fid)
```
Comment (chrislusf) [inline go/operation/chunked_file.go]: Print the "master" value in the log.

File: go/weed/weed_server/volume_server_handlers_read.go [hunk 3]
```diff
@@ -215,3 +220,145 @@ func (vs *VolumeServer) GetOrHeadHandler(w http.ResponseWriter, r *http.Request)
...
+	rangeReq := r.Header.Get("Range")
+	if rangeReq == "" {
+		w.Header().Set("Content-Length", strconv.FormatInt(chunkManifest.Size, 10))
+		if _, e = io.Copy(w, chunkedFileReader); e != nil {
+			glog.V(2).Infoln("response write error:", e)
+		}
+		return true
+	}
+
+	//the rest is dealing with partial content request
```
Comment (chrislusf) [inline go/weed/weed_server/volume_server_handlers_read.go:201]: Can the following logic for "partial content request" go into a function call, so we can share the code in the non-chunked reading code path also?

**Comment by chrislusf** [comment]
Thanks a lot for your impressive work! This seems also affects the command "weed download". Please check it also.

**Comment by tnextday** [comment]
Hi chrislusf, I have update the code, please check it again.

**Comment by chrislusf** [comment]
Thanks a lot!

I added some format changes after merged this commit. Maybe you can sync the code before working on other features.

**Comment by tnextday** [comment]
OK.

**Comment by chrislusf** [comment]
Hi, tnextday, this pull request is very nice! Could you consider adding an a POST call to append to a file?

**Pull request merged**